### PR TITLE
Add description of return removal

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/remove_returns.h [hunk 1]
```diff
@@ -11,6 +11,74 @@ Date:   September 2009
...
+/// are set to void as well (represented by the type `empty_typet`). This
+/// applies both to the functions (i.e., the member `type` of `goto_functiont`)
+/// and to the call sites (i.e., the type
+/// `to_code_function_call(code).function().type()` with `code` being the code
+/// member of the `instructiont` instance that represents the function call).
+///
+/// The types of function pointer expressions in the goto program are however
+/// not changed. For example, in an assignment where `func` is assigned to a
+/// function pointer, such as `int (*f)(void) = func`, the function types
+/// appearing in the lhs and rhs both retain the integer return type.
```
Comment (danpoe) [inline src/goto-programs/remove_returns.h:80]: This paragraph documents the current behavior which should probably be changed at some point. Not sure if there would be any pitfalls with it though.

File: src/goto-programs/remove_returns.h [hunk 2]
```diff
@@ -11,6 +11,74 @@ Date:   September 2009
 /// \file
 /// Remove function returns
 
+/// The functions `remove_returns()` remove return instructions from a goto
+/// program. The `restore_returns()` functions restore the returns in a goto
+/// program in which they have been removed previously.
+///
+/// In a goto program in which returns have not beem removed, functions return
```
Comment (peterschrammel) [inline src/goto-programs/remove_returns.h]: been

File: src/goto-programs/remove_returns.h [hunk 3]
```diff
@@ -11,6 +11,74 @@ Date:   September 2009
...
+///    ...
+/// 1: END_FUNCTION
+///
+/// func();
+/// r = func#return_value;
+///
+/// ```
+///
+/// As `return` instructions are removed, the return types of the function types
+/// are set to void as well (represented by the type `empty_typet`). This
```
Comment (peterschrammel) [inline src/goto-programs/remove_returns.h:71]: What's the difference between `void_typet` and `empty_typet` ?

**Review by peterschrammel — APPROVED** [review]

File: src/goto-programs/remove_returns.h (hunk 3 above)
Comment (danpoe) [reply src/goto-programs/remove_returns.h:71]: At the moment `void_typet` is defined as `class void_typet:public empty_typet {};`. I think use of `void_typet` is deprecated now.

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 36bff33df522bcad27c10d71d3b00ef5a8c38127).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99812467

File: src/goto-programs/remove_returns.cpp [hunk 4]
```diff
@@ -9,7 +9,7 @@ Date:   September 2009
 \*******************************************************************/
 
 /// \file
-/// Remove function return values
+/// Remove function returns
```
Comment (tautschnig) [inline src/goto-programs/remove_returns.cpp]: Maybe "Replace function returns by assignments to global variables"

File: src/goto-programs/remove_returns.h [hunk 5]
```diff
@@ -11,6 +11,74 @@ Date:   September 2009
...
+///    ...
+///    return 1;
+///    GOTO 1
+///    ...
+/// 1: END_FUNCTION
+///
+/// r = func();
+/// ```
+///
+/// The return removal pass introduces a global variable `func#return_value`
```
Comment (tautschnig) [inline src/goto-programs/remove_returns.h]: thread-local global variable

File: src/goto-programs/remove_returns.h (hunk 3 above)
Comment (tautschnig) [reply src/goto-programs/remove_returns.h:71]: Yes, I really think we should get rid of `void_typet` - we don't use it consistently and effectively rely on `empty_typet() == void_typet()`. I'll put up a PR.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 39b453a1ff20a972f36b6ad0768924df4507be37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99918689

**Pull request merged**